Add a demo file

Gjacquenot opened the pull request:
This pull request does several things

- it adds a demo based on the data proposed in the README.md
- it includes an image of the demo result
- it adds a Dockefile to create a complete compilation env for the project
- it add .travis.yml file that enables travis compilation with docker. Log of this PR can be seen [here](https://travis-ci.org/github/Gjacquenot/decomp/builds/671069656)

If accepted, one will have to update the URL of the travis badge to match the maintainer associated account.

ltjax commented:
Hey there! Wow, that looks like a lot of work, thanks for that. A few things before I review this fully:
Did you see the functions for generating output in output.hpp? From what I saw, you basically wrote a python code generator.
I'm not a big fan of using Docker within a freshly provisioned VM, like is the case with travis. It kinda defeats the point, unless you really want to deploy a container or the VM cannot give you what you need. I could just merge this and then port it to use a more lean conan environment. I meant to do that anyways.
I assume you're using the Makefile for testing your CMake run? I'd rather not have that in the repository, it suggests that you can build this with both CMake and make, while it's really something totally different.
Again, don't wanna discourage you, it's awesome that you want to contribute. Thanks again!

Gjacquenot commented:
Hello @ltjax!

Since you added properly the travis compilation with conan, I have removed my contribution.
I have removed the Makefile to  avoid any confusion with CMakeLists.txt

I can also remove the Dockerfile if wanted.

I have updated the demo.cpp file to export input and output to SVG: I guess it can be enhanced since I had to scale input to obtain nice SVG, and I had to give output as a list of holes in the writePolygon function.

I can enhance the python script generation but need some advice on how to update output.cpp and output.hpp.

ltjax commented:
Ah, I was thinking you could replace the "output via python" with output to SVG. But apparently you're doing something else there as well. I'll have to read the code thoroughly and get back to you on that!

ltjax commented:
Okay, so I've looked at this in more detail, and it looks indeed like you are basically using matplotlib to render the input and output meshes, with some random colors (I like the random seed, it's metal).
From what I gather, the difference between what the python script does and the svg is basically that that the python script renders "nicer", i.e. with nice "auto" scaling and colors.
So we have a few options:
* Make the svg prettier. Adding color support and scaling should be pretty easy. I'm not so sure about the grid, the box and labels. I guess matplotlib rules at this stuff, which brings me to my other idea:
* Instead of the python code generator, write a very simple json output (just using ostream), and turn the python code into a real .py file that can read that json and render it like you're doing now. This should not be more or much different code than you have now, it's just separated a lot more nicely. JSON output is something I can totally get behind for putting in the library (can be in output.cpp/output.hpp) , and a separate script to show it would be great.

Gjacquenot commented:
The second Idea seems nice. I like the separation you propose . All we have to do is to find a json format.

I am pretty sûre that someone has already proposed one.

Or the standard de facto for input and output is the poly file format ( old but already défined)
It is used here https://github.com/Gjacquenot/polypartition

ltjax commented:
I wouldn't worry so much about the format for now and focus on simplicity instead. basically just dump the vertices and indices into any kind of file. The poly file format is fine with me, as well. My main point is to use something that is easy to write from c++, and easy to read from python.

Gjacquenot commented:
I would like to propose something like this with a simple C++ function and the associated JSON output. @ltjax , is it ok with you? The proposed JSON format could be used as input with fields `vertices` and `input`.

```cpp
std::ostream & dump_json(
    std::ostream &out,
    const std::vector<decomp::Point>& pointList,
    const std::vector<std::uint16_t>& outerPolygon,
    const std::vector<std::vector<std::uint16_t>>& holeList,
    const std::vector<std::vector<std::uint16_t>>& convexPolygonList)
{
    out << "{" << std::endl;
    out << "\"vertices\": [" << std::endl;
    for (auto& p:pointList) out << "[" << p[0] << "," << p[1] << "],";
    out << '\b';
    out << "]," << std::endl;
    out << "\"input\": {" << std::endl;
    out << "\"outer\": " << std::endl << "[";
    for (auto& p:outerPolygon) out << p << ",";
    out << "\b]," << std::endl;
    out << "\"holes\": " << std::endl << "[";

    for (auto& hole:holeList)
    {
        out << "[";
        for (auto& e:hole)
        {
            out << e << ",";
        }
        if (not hole.empty()) out << "\b";
        out << "],";
    }
    if (not holeList.empty()) out << "\b";
    out << "]}," << std::endl;
    out << "\"output\":" << std::endl << "[";
    for (auto& convexPolygon:convexPolygonList)
    {
        out << "[";
        for (auto& e:convexPolygon)
        {
            out << e << ",";
        }
        if (not convexPolygon.empty()) out << "\b";
        out << "],";
    }
    if (not convexPolygonList.empty()) out << "\b";
    out << "]" << std::endl;
    out << "}" << std::endl;
    return out;
}
```


```json
{
"vertices": [
[-4,0],[-3,-2],[3,-2],[4,0],[3,2],[-3,2],[-3,0],[-2,-1],[-1,0],[-2,1],[1,0],[2,-1],[3,0],[2,1]],
"input": {
"outer": 
[0,1,2,3,4,5],
"holes": 
[[13,12,11,10],[9,8,7,6]]},
"output":
[[3,4,12,2],[13,12,4],[2,12,11],[8,10,13,4,5,9],[11,10,8,7,1,2],[6,9,5],[6,5,0,1],[7,6,1]]
}
```

ltjax commented:
Hey @Gjacquenot . The output is perfect, we can leave it like that. I see the code is a little too complex for what it does - and it seems like a lot of extra complexity comes from removing that trailing comma in the arrays, especially using `\b` is pretty hacky. How about using a small tool function that solves this problem once, like this:
```
template <class Container, class F>
void writeJsonArray(std::ostream& out, Container const& container, F f)
{
  out << "[";
  auto i = container.begin();
  if (i != container.end())
    out << f(out, *i++);
  for (;i != container.end();++i)
    out << "," << f(out, *i)
  out << "]";
}
```
For the holes, which I think are the most complex case, use it like:

```
writeJsonArray(out, holeList, [](std::ostream& out, decomp::IndexList const& hole)
{
  writeJsonArray(out, hole, [](std::ostream& out, std::uint16_t index)
  {
    out << index;
  });
});
```

Disclaimer: did not try, just typed it here. I hope the idea is clear.

Gjacquenot commented:
@ltjax I have added a Python script to create image from JSON output and I have used your templated syntax to write JSON array. I had to fix some typos in your code, but code is neater!
Maybe we can move this code in the output.cpp file?

```cpp

using namespace decomp;

template <class Container, class F>
void writeJsonArray(std::ostream& out, Container const& container, F f)
{
  out << "[";
  auto i = container.begin();
  if (i != container.end())
  f(out, *i++);
  for (;i != container.end();++i)
  {
    out << ",";
    f(out, *i);
  }
  out << "]";
}

std::ostream & dump_json(
    std::ostream &out,
    const PointList& pointList,
    const IndexList& outerPolygon,
    const std::vector<IndexList>& holeList,
    const std::vector<IndexList>& convexPolygonList)
{
  out << "{" << std::endl;
  out << "\"vertices\":" << std::endl;
  writeJsonArray(out, pointList, [](std::ostream& os, const Point& p) {os<<"["<<p[0]<<","<<p[1]<<"]";});
  out << "," << std::endl;
  out << "\"input\": {" << std::endl;
  out << "\"outer\": " << std::endl;
  writeJsonArray(out, outerPolygon, [](std::ostream& os, const std::uint16_t& p) {os << p;});
  out << "," << std::endl;
  out << "\"holes\": " << std::endl;
  writeJsonArray(out, holeList, [](std::ostream& os, IndexList const& hole)
  {
    writeJsonArray(os, hole, [](std::ostream& os, std::uint16_t index){os << index;});
  });
  out << std::endl << "}," << std::endl;
  out << "\"output\":" << std::endl << "[";
  writeJsonArray(out, convexPolygonList, [](std::ostream& os, IndexList const& hole)
  {
    writeJsonArray(os, hole, [](std::ostream& os, std::uint16_t index){os << index;});
  });
  out << "}" << std::endl;
  return out;
}
```

ltjax commented on CMakeLists.txt:
```diff
@@ -60,4 +60,16 @@ if(${PROJECT_NAME}_BUILD_TESTS)
   endif()
+  include(CTest)
+  enable_testing()
+  add_test(NAME test_syntaxe COMMAND test_${PROJECT_NAME})
```
test_syntaxe doesn't mean anything to me. Can you explain or just call it test_decomp?

Gjacquenot replied:
I have renamed the test with the name of the test program `test_${PROJECT_NAME}`

ltjax commented on README.md:
```diff
@@ -1,6 +1,7 @@
 # decomp
 
 [![Build status](https://ci.appveyor.com/api/projects/status/485fudsl4vv5ikfy?svg=true)](https://ci.appveyor.com/project/ltjax/decomp)
+[![Build Status](https://travis-ci.org/Gjacquenot/decomp.svg?branch=master)](https://travis-ci.org/Gjacquenot/decomp)
```
You can use mine: https://travis-ci.org/ltjax/decomp.svg?branch=master

ltjax commented on demo/demo.cpp:
```diff
@@ -0,0 +1,90 @@
+  out << "[";
+  auto i = container.begin();
+  if (i != container.end())
+  f(out, *i++);
```
Prolly want some indentation here

ltjax commented on demo/demo.cpp:
```diff
@@ -0,0 +1,90 @@
+  out << "]";
+}
+
+std::ostream & dump_json(
```
Rest of this code base uses camelCase, so I'd use that here, too

ltjax commented on demo/plot_decomp.py:
```diff
@@ -0,0 +1,42 @@
+import numpy as np
+import random
+import sys
+random.seed(666)
```
:metal:

ltjax commented:
Yes, feel free to incorporate this into the output component! Looks great so far!

Gjacquenot commented:
I will take your.comments into account tomorrow!

Le lun. 13 avr. 2020 à 01:04, Marius Elvert <notifications@github.com> a
écrit :

> *@ltjax* requested changes on this pull request.
> ------------------------------
>
> In CMakeLists.txt
> <https://github.com/ltjax/decomp/pull/6#discussion_r407262817>:
>
> > @@ -60,4 +60,16 @@ if(${PROJECT_NAME}_BUILD_TESTS)
>
>      target_link_libraries(${TEST_NAME}
>
>        PUBLIC decomp Catch2::Catch2)
>
>    endif()
>
> +  include(CTest)
>
> +  enable_testing()
>
> +  add_test(NAME test_syntaxe COMMAND test_${PROJECT_NAME})
>
>
> test_syntaxe doesn't mean anything to me. Can you explain or just call it
> test_decomp?
> ------------------------------
>
> In README.md
> <https://github.com/ltjax/decomp/pull/6#discussion_r407262968>:
>
> > @@ -1,6 +1,7 @@
>
>  # decomp
>
>
>
>  [![Build status](https://ci.appveyor.com/api/projects/status/485fudsl4vv5ikfy?svg=true)](https://ci.appveyor.com/project/ltjax/decomp)
>
> +[![Build Status](https://travis-ci.org/Gjacquenot/decomp.svg?branch=master)](https://travis-ci.org/Gjacquenot/decomp)
>
>
> You can use mine: https://travis-ci.org/ltjax/decomp.svg?branch=master
> ------------------------------
>
> In demo/demo.cpp
> <https://github.com/ltjax/decomp/pull/6#discussion_r407263110>:
>
> > +#include <iostream>
>
> +
>
> +#include <decomp/convex_decomposition.hpp>
>
> +#include <decomp/triangulation.hpp>
>
> +#include <decomp/operations.hpp>
>
> +#include <decomp/output.hpp>
>
> +
>
> +using namespace decomp;
>
> +
>
> +template <class Container, class F>
>
> +void writeJsonArray(std::ostream& out, Container const& container, F f)
>
> +{
>
> +  out << "[";
>
> +  auto i = container.begin();
>
> +  if (i != container.end())
>
> +  f(out, *i++);
>
>
> Prolly want some indentation here
> ------------------------------
>
> In demo/demo.cpp
> <https://github.com/ltjax/decomp/pull/6#discussion_r407263225>:
>
> > +template <class Container, class F>
>
> +void writeJsonArray(std::ostream& out, Container const& container, F f)
>
> +{
>
> +  out << "[";
>
> +  auto i = container.begin();
>
> +  if (i != container.end())
>
> +  f(out, *i++);
>
> +  for (;i != container.end();++i)
>
> +  {
>
> +    out << ",";
>
> +    f(out, *i);
>
> +  }
>
> +  out << "]";
>
> +}
>
> +
>
> +std::ostream & dump_json(
>
>
> Rest of this code base uses camelCase, so I'd use that here, too
> ------------------------------
>
> In demo/plot_decomp.py
> <https://github.com/ltjax/decomp/pull/6#discussion_r407263354>:
>
> > @@ -0,0 +1,42 @@
>
> +
>
> +"""
>
> +This script reads JSON input from stdin and creates a matplotlib figure
>
> +showing input and output of decomp program.
>
> +"""
>
> +import json
>
> +import matplotlib.pyplot as plt
>
> +import numpy as np
>
> +import random
>
> +import sys
>
> +random.seed(666)
>
>
> 🤘
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ltjax/decomp/pull/6#pullrequestreview-391902049>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABT3QHFE5XIRYGJC7IFKA2LRMJCJNANCNFSM4L7GXLGA>
> .
>

Gjacquenot commented:
I think we are good now. Two things that can be done:

- The python script to plot input and output is not installed. We could update the CMakeLists.txt with a `install(files` command?
- I can squash some commits to have a neater git history.

ltjax commented:
> 
> 
> I think we are good now. Two things that can be done:

> 
>     * The python script to plot input and output is not installed. We could update the CMakeLists.txt with a `install(files` command?
Nah, I think that's more for people playing around with the lib rather than pure consumers.
> 
>     * I can squash some commits to have a neater git history.
I prefer having the real, messy history! Every commit matters.

ltjax commented:
For the record, there are some style inconsistencies (ref placement, east-const and identation) but that can be fixed whenever.

ltjax commented:
Thanks very much for the contribution!

Gjacquenot commented:
@ltjax Thank you!
